OpenRA.Game/ExternalMods.cs

```diff
@@ -107,53 +113,96 @@ internal void Register(Manifest mod)
-			{
-				Log.Write("debug", "Failed to register currrent mod metadata");
-				Log.Write("debug", e.ToString());
-			}
+			// Make sure the mod is available for this session, even if saving it fails
+			LoadMod(yaml.First().Value);
 
-			// Clean up stale mod registrations:
-			//  - LaunchPath no longer exists (uninstalled)
-			//  - LaunchPath and mod match the current mod, but version differs (newer version installed on top)
-			List<string> toRemove = null;
-			foreach (var kv in mods)
+			foreach (var source in locations)
 			{
-				var k = kv.Key;
-				var m = kv.Value;
-				if (!File.Exists(m.LaunchPath) || (m.LaunchPath == launchPath && m.Id == mod.Id && k != key))
+				if (!Directory.Exists(source))
+					continue;
+
+				var metadataPath = Path.Combine(source, "ModMetadata");
+
+				try
+				{
+					Directory.CreateDirectory(metadataPath);
+					File.WriteAllLines(Path.Combine(metadataPath, key + ".yaml"), yaml.ToLines(false).ToArray());
+				}
+				catch (Exception e)
 				{
-					Log.Write("debug", "Removing stale mod metadata entry '{0}'", k);
-					if (toRemove == null)
-						toRemove = new List<string>();
+					Log.Write("debug", "Failed to register current mod metadata");
+					Log.Write("debug", e.ToString());
+				}
 
-					toRemove.Add(k);
-					var path = Path.Combine(supportPath, k + ".yaml");
-					try
+				// Clean up stale mod registrations:
+				//  - LaunchPath no longer exists (uninstalled)
+				//  - LaunchPath and mod match the current mod, but version differs (newer version installed on top)
+				List<string> toRemove = null;
+				foreach (var kv in mods)
+				{
+					var k = kv.Key;
+					var m = kv.Value;
+					if (!File.Exists(m.LaunchPath) || (m.LaunchPath == launchPath && m.Id == mod.Id && k != key))
 					{
-						File.Delete(path);
-					}
-					catch (Exception e)
-					{
-						Log.Write("debug", "Failed to remove mod metadata file '{0}'", path);
-						Log.Write("debug", e.ToString());
+						Log.Write("debug", "Removing stale mod metadata entry '{0}'", k);
+						if (toRemove == null)
+							toRemove = new List<string>();
+
+						toRemove.Add(k);
+						var path = Path.Combine(metadataPath, k + ".yaml");
+						try
+						{
+							File.Delete(path);
+						}
+						catch (Exception e)
+						{
+							Log.Write("debug", "Failed to remove mod metadata file '{0}'", path);
+							Log.Write("debug", e.ToString());
+						}
 					}
 				}
+
+				if (toRemove != null)
+					foreach (var r in toRemove)
+						mods.Remove(r);
 			}
+		}
+
+		internal void Unregister(Manifest mod, ModRegistration registration)
+		{
+			var locations = new List<string>();
+			if (registration.HasFlag(ModRegistration.System))
+				locations.Add(Platform.SystemSupportDir);
+
+			if (registration.HasFlag(ModRegistration.User))
+				locations.Add(Platform.SupportDir);
 
-			if (toRemove != null)
-				foreach (var r in toRemove)
-					mods.Remove(r);
+			var key = ExternalMod.MakeKey(mod);
+			mods.Remove(key);
```

Comment:
This simply returns false if the mod doesn't exist.  The only way it can crash is if `key` is null, which isn't possible.